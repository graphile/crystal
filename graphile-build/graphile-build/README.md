# graphile-build

[![GitHub Sponsors](https://img.shields.io/github/sponsors/benjie?color=ff69b4&label=github%20sponsors)](https://github.com/sponsors/benjie)
[![Patreon sponsor button](https://img.shields.io/badge/sponsor-via%20Patreon-orange.svg)](https://patreon.com/benjie)
[![Discord chat room](https://img.shields.io/discord/489127045289476126.svg)](http://discord.gg/graphile)
[![Follow](https://img.shields.io/badge/twitter-@GraphileHQ-blue.svg)](https://twitter.com/GraphileHQ)

`graphile-build` provides a framework to build extensible GraphQL APIs by
combining plugins. Each plugin typically has its own small purpose (such as
implementing the [`Node` interface][global-object-identification], adding
`query: Query` to mutation payloads, or watching an external source for schema
changes) and by combining these plugins together you get a large, powerful, and
manageable GraphQL schema. Plugins enable you to make broad changes to your
GraphQL schema with minimal code, and because you're changing your schema in
broad ways it helps to guarantee consistency.

Graphile Build has strong built-in support for [Gra*fast*](https://grafast.org)
so you can make an exceptionally performant auto-generated (or
generator-assisted) GraphQL API.

[PostGraphile](https://github.com/graphile/postgraphile) uses Graphile Build and
Gra*fast* to produce an extremely high performance standards-compliant GraphQL
API with minimal fuss. The separate
[`graphile-build-pg` module](../graphile-build-pg/README.md) contains the
plugins that are specific to PostgreSQL support (`graphile-build` itself does
not know nor care about databases).

<!-- SPONSORS_BEGIN -->

## Crowd-funded open-source software

To help us develop this software sustainably, we ask all individuals and
businesses that use it to help support its ongoing maintenance and development
via sponsorship.

### [Click here to find out more about sponsors and sponsorship.](https://www.graphile.org/sponsor/)

And please give some love to our featured sponsors ðŸ¤©:

<table><tr>
<td align="center"><a href="https://www.the-guild.dev/"><img src="https://graphile.org/images/sponsors/theguild.png" width="90" height="90" alt="The Guild" /><br />The Guild</a> *</td>
<td align="center"><a href="https://dovetailapp.com/"><img src="https://graphile.org/images/sponsors/dovetail.png" width="90" height="90" alt="Dovetail" /><br />Dovetail</a> *</td>
<td align="center"><a href="https://www.netflix.com/"><img src="https://graphile.org/images/sponsors/Netflix.png" width="90" height="90" alt="Netflix" /><br />Netflix</a> *</td>
<td align="center"><a href="https://stellate.co/"><img src="https://graphile.org/images/sponsors/Stellate.png" width="90" height="90" alt="Stellate" /><br />Stellate</a> *</td>
</tr><tr>
<td align="center"><a href="https://www.accenture.com/"><img src="https://graphile.org/images/sponsors/accenture.svg" width="90" height="90" alt="Accenture" /><br />Accenture</a> *</td>
<td align="center"><a href="https://microteam.io/"><img src="https://graphile.org/images/sponsors/micro.png" width="90" height="90" alt="We Love Micro" /><br />We Love Micro</a> *</td>
</tr></table>

<em>\* Sponsors the entire Graphile suite</em>

<!-- SPONSORS_END -->

## Documentation

TODO!

[global-object-identification]:
  https://relay.dev/graphql/objectidentification.htm
