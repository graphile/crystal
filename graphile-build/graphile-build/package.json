{
  "name": "graphile-build",
  "version": "5.0.0-beta.33",
  "description": "Build a GraphQL schema from plugins",
  "type": "commonjs",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "scripts": {
    "prepack": "tsc -b && cp src/.npmignore dist/.npmignore"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graphile/crystal.git"
  },
  "keywords": [
    "graphile",
    "graphql",
    "engine",
    "apollo",
    "graphqljs",
    "plugin",
    "build",
    "extension",
    "graphite"
  ],
  "author": "Benjie Gillam <code@benjiegillam.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/graphile/crystal/issues"
  },
  "homepage": "https://graphile.org/graphile-build/",
  "dependencies": {
    "@types/node": "^20.5.7",
    "@types/pluralize": "^0.0.30",
    "@types/semver": "^7.5.1",
    "chalk": "^4.1.2",
    "debug": "^4.3.4",
    "graphile-config": "workspace:^",
    "graphql": "^16.1.0-experimental-stream-defer.6",
    "lodash": "^4.17.21",
    "pluralize": "^7.0.0",
    "semver": "^7.5.4",
    "tamedevil": "workspace:^",
    "tslib": "^2.6.2"
  },
  "engines": {
    "node": ">=16"
  },
  "peerDependencies": {
    "grafast": "workspace:^",
    "graphile-config": "workspace:^",
    "graphql": "^16.1.0-experimental-stream-defer.6"
  },
  "files": [
    "dist",
    "index.js"
  ],
  "devDependencies": {
    "@types/debug": "^4.1.8",
    "@types/jest": "^29.5.4",
    "graphile-export": "workspace:^",
    "graphql": "16.1.0-experimental-stream-defer.6",
    "jest": "^29.6.4",
    "ts-node": "^10.9.1",
    "typescript": "^5.2.2"
  }
}
