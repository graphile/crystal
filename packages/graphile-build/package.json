{
  "name": "graphile-build",
  "version": "4.7.0",
  "description": "Build a GraphQL schema from plugins",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "jest -i",
    "prepack": "tsc -b"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/graphile/graphile-build.git"
  },
  "keywords": [
    "graphile",
    "graphql",
    "engine",
    "apollo",
    "graphqljs",
    "plugin",
    "build",
    "extension"
  ],
  "author": "Benjie Gillam <code@benjiegillam.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/graphile/graphile-build/issues"
  },
  "homepage": "https://graphile.org/graphile-build/",
  "dependencies": {
    "@graphile/lru": "4.6.0-alpha.0",
    "@types/node": "^14.0.1",
    "@types/pluralize": "^0.0.29",
    "@types/semver": "^7.2.0",
    "chalk": "^4.0.0",
    "debug": "^4.2.0",
    "graphql-parse-resolve-info": "4.7.0",
    "iterall": "^1.3.0",
    "lodash": "^4.17.15",
    "pluralize": "^7.0.0",
    "semver": "^7.3.2",
    "tslib": "^2.0.0"
  },
  "engines": {
    "node": ">=12.16"
  },
  "peerDependencies": {
    "graphql": "^14.6.0"
  },
  "files": [
    "dist",
    "index.js"
  ],
  "devDependencies": {
    "@types/debug": "^4.1.5",
    "@types/jest": "^25.2.3",
    "@types/lodash": "^4.14.151",
    "graphql": "^14.6.0",
    "jest": "^26.0.1",
    "ts-node": "^8.10.1",
    "typescript": "^3.9.2"
  }
}
