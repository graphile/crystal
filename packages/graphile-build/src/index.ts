import "./global";
import util from "util";
import SchemaBuilder from "./SchemaBuilder";
import {
  SwallowErrorsPlugin,
  StandardTypesPlugin,
  NodePlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  ClientMutationIdDescriptionPlugin,
  MutationPayloadQueryPlugin,
  AddQueriesToSubscriptionsPlugin,
} from "./plugins";
import resolveNode from "./resolveNode";
export { GetDataFromParsedResolveInfoFragmentFunction } from "./makeNewBuild";

export {
  constantCaseAll,
  formatInsideUnderscores,
  upperFirst,
  camelCase,
  constantCase,
  upperCamelCase,
  pluralize,
  singularize,
} from "./utils";

export { SchemaBuilder };

export { LiveSource, LiveProvider, LiveMonitor, LiveCoordinator } from "./Live";

export const getBuilder = async (
  plugins: Array<GraphileEngine.Plugin>,
  options: GraphileEngine.GraphileBuildOptions = {},
): Promise<SchemaBuilder> => {
  const builder = new SchemaBuilder(options);
  for (let i = 0, l = plugins.length; i < l; i++) {
    const plugin = plugins[i];
    if (typeof plugin !== "function") {
      throw new Error(
        `Expected a list of plugin functions, instead list contained a non-function at index ${i}: ${util.inspect(
          plugin,
        )}`,
      );
    }
    // $FlowFixMe: displayName
    builder._setPluginName(plugin.displayName || plugin.name);
    await plugin(builder, options);
    builder._setPluginName(null);
  }
  return builder;
};

export const buildSchema = async (
  plugins: Array<GraphileEngine.Plugin>,
  options: GraphileEngine.GraphileBuildOptions = {},
): Promise<import("graphql").GraphQLSchema> => {
  const builder: SchemaBuilder = await getBuilder(plugins, options);
  return builder.buildSchema();
};

export const defaultPlugins: Array<GraphileEngine.Plugin> = [
  SwallowErrorsPlugin,
  StandardTypesPlugin,
  NodePlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  ClientMutationIdDescriptionPlugin,
  MutationPayloadQueryPlugin,
  AddQueriesToSubscriptionsPlugin,
];

export {
  SwallowErrorsPlugin,
  StandardTypesPlugin,
  NodePlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  ClientMutationIdDescriptionPlugin,
  MutationPayloadQueryPlugin,
  AddQueriesToSubscriptionsPlugin,
  // resolveNode: EXPERIMENTAL, API might change!
  resolveNode,
};
