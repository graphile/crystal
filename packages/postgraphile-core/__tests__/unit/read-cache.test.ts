jest.mock("graphile-build");
jest.setTimeout(20000);

import mockFs from "mock-fs";
import * as graphileBuild from "graphile-build";
import { getPostGraphileBuilder } from "../..";

type MockHack<T extends (...args: any[]) => any> = jest.Mock<ReturnType<T>>;
const mockedGetBuilder = graphileBuild.getBuilder as MockHack<
  typeof graphileBuild.getBuilder
>;

beforeEach(() => {
  // required for mocks assertions
  jest.resetAllMocks();
  mockFs.restore();
});

afterEach(() => {
  mockFs.restore();
});

/*
Current behaviour of getPostGraphileBuilder and readCache option:

If readCache is String: 
1. cacheString is loaded by reading the file defined by readCache
2. memoizeCache is set by JSON.parse(cacheString)

If readCache is Object:
1. memoizeCache is set as readCache directly

Then
3. persistentMemoizeWithKey is a function created dynamically that will use memoizeCache
4. getBuilder(imported from graphile-build) is called, passing persistentMemoizeWithKey as one of the arguments

Test strategy for readCache

1. Mock fs.readFile to control input
2. Mock getBuilder to control output
3. Test different inputs and validate that getBuilder is called with expected output
*/
test("when no readCache flag, persistentMemoizeWithKey should be undefined", async () => {
  // mock getBuilder to fake output
  const expectedOutput = {};
  mockedGetBuilder.mockResolvedValueOnce(expectedOutput as any);
  // call our method and test output
  const output = await getPostGraphileBuilder({} as any, [], {});
  expect(output).toBe(expectedOutput);
  expect(graphileBuild.getBuilder).toHaveBeenCalledTimes(1);
  // check persistentMemoizeWithKey, the actual "result" of readCache flag
  const { persistentMemoizeWithKey } = mockedGetBuilder.mock.calls[0][1];
  expect(persistentMemoizeWithKey).toBeUndefined();
});

describe("When readCache is String", () => {
  test("if cache file has content, persistentMemoizeWithKey should be a valid function", async () => {
    mockFs({
      "path/to/cache": '{ "__test": true }',
    });

    // mock getBuilder to fake output
    const expectedOutput = {};
    mockedGetBuilder.mockResolvedValueOnce(expectedOutput as any);
    // call our method and test output
    const output = await getPostGraphileBuilder({} as any, [], {
      readCache: "path/to/cache",
    });
    mockFs.restore();
    expect(output).toBe(expectedOutput);
    expect(mockedGetBuilder).toHaveBeenCalledTimes(1);
    // check persistentMemoizeWithKey, the actual "result" of readCache flag
    const { persistentMemoizeWithKey } = mockedGetBuilder.mock.calls[0][1];
    expect(typeof persistentMemoizeWithKey).toBe("function");
    expect(persistentMemoizeWithKey("__test")).toEqual(true);
    expect(() => persistentMemoizeWithKey("unknown_key")).toThrow();
  });

  test("if cache file has invalid content, getPostGraphileBuilder should error", async () => {
    mockFs({
      "path/to/cache": "thisisnotjson",
    });

    // call our method and check error

    let error;
    try {
      await getPostGraphileBuilder({} as any, [], {
        readCache: "path/to/cache",
      });
    } catch (e) {
      error = e;
    }
    mockFs.restore();
    expect(error).toBeDefined();
    expect(error).toMatchSnapshot();
  });
});

describe("When readCache is Object", () => {
  test("persistentMemoizeWithKey should be a valid function", async () => {
    // mock getBuilder to fake output
    const expectedOutput = {};
    mockedGetBuilder.mockResolvedValueOnce(expectedOutput as any);
    // call our method and test output
    const output = await getPostGraphileBuilder({} as any, [], {
      readCache: { __test: true },
    });
    expect(output).toBe(expectedOutput);
    expect(mockedGetBuilder).toHaveBeenCalledTimes(1);
    // check persistentMemoizeWithKey, the actual "result" of readCache flag
    const { persistentMemoizeWithKey } = mockedGetBuilder.mock.calls[0][1];
    expect(typeof persistentMemoizeWithKey).toBe("function");
    expect(persistentMemoizeWithKey("__test")).toEqual(true);
    expect(() => persistentMemoizeWithKey("unknown_key")).toThrow();
  });
});

describe("when readCache is not String or Object, getPostGraphileBuilder should error", () => {
  test("when its Boolean", async () => {
    // call our method with invalid readCache value and check error
    let error;
    try {
      await getPostGraphileBuilder({} as any, [], {
        readCache: true as any,
      });
    } catch (e) {
      error = e;
    }
    expect(error).toBeDefined();
    mockFs.restore();
    expect(error).toMatchSnapshot();
  });
  test("when its Array", async () => {
    // call our method with invalid readCache value and check error
    let error;
    try {
      await getPostGraphileBuilder({} as any, [], {
        readCache: [],
      });
    } catch (e) {
      error = e;
    }
    expect(error).toBeDefined();
    mockFs.restore();
    expect(error).toMatchSnapshot();
  });
  test("when its Number", async () => {
    // call our method with invalid readCache value and check error
    let error;
    try {
      await getPostGraphileBuilder({} as any, [], {
        readCache: 3 as any,
      });
    } catch (e) {
      error = e;
    }
    expect(error).toBeDefined();
    mockFs.restore();
    expect(error).toMatchSnapshot();
  });
});
