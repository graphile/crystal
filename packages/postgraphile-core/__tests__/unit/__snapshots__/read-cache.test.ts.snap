// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When readCache is String if cache file has invalid content, getPostGraphileBuilder should error 1`] = `[Error: Failed to parse cache file 'path/to/cache', perhaps it is corrupted? SyntaxError: Unexpected token h in JSON at position 1]`;

exports[`when readCache is not String or Object, getPostGraphileBuilder should error when its Array 1`] = `[Error: 'readCache' not understood; expected string or object, but received 'array']`;

exports[`when readCache is not String or Object, getPostGraphileBuilder should error when its Boolean 1`] = `[Error: 'readCache' not understood; expected string or object, but received 'boolean']`;

exports[`when readCache is not String or Object, getPostGraphileBuilder should error when its Number 1`] = `[Error: 'readCache' not understood; expected string or object, but received 'number']`;
